[skip ci] voz/fsdp_autograd3 tracker PR

ezyang opened the pull request:
cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @kiukchung @d4l3k @lucasllc @Xia-Weiwen

voznesenskym committed:
[Dynamo x FSDP][2/x] Small changes to distributed to make it dynamo friendly

ghstack-source-id: 356fab194dcb61fdd12ae02a5b2ce67e1cf46d69
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106886

[Dynamo x FSDP][2/x] Small changes to distributed to make it dynamo friendly

Dist nits

voznesenskym committed:
[Dynamo x FSDP][3/x] TypedStorage and storage_offset

ghstack-source-id: caf0a99c5ea9558549f15a1da5701122212646b1
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106888

Fix

Better modules

voznesenskym committed:
[Dynamo x FSDP][1/x] Builder support for deque

ghstack-source-id: 0d52cb324250e98c7d399300a938b4d43fc8f293
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106884

Appendleft test

voznesenskym committed:
[Dynamo x FSDP][5/x] Fix bug in __class__ sources

ghstack-source-id: 828ec43504d249b4f51760d2245123c5a69cb7a6
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106906

voznesenskym committed:
[Dynamo x FSDP][4/x] Add support for params, buffers, submodules on FSDPManagedNNModuleVariable

fix2

ghstack-source-id: d50fb1b2843b4031a852db2b2d8a34eb7b941806
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106890

fix2

fix2

Lint

Lint

voznesenskym committed:
[FSDP][WIP] Trace FSDP

[FSDP][6/N] Add util to check valid param freezing

Set initial

Set initial

wip

WIP

WIP

WIP

lint

lint

lint

lint

lint

Fix

Fix

fixup

fixup

Fix

Initial skipfiles

WIP

WIP

WIP

[dynamo x fsdp] Simplify stream logic handling

WIP

[dynamo x fsdp] Remove uncessary partials

[dynamo x fsdp] Remove uncessary partials

Lint

WIP

fixup

fixup

fixup

fixup

fixup

Storage_offset

Storage_offset

Storage_offset

Fix

Fix

Fixes

Fixes

Fixes

Fixes

wip

wip

wip

wip

wip

wip

wip

wip

wip

Undo

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

wip

wip

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

wip

wip

wip

wip

wip

wip

wip

wip

wip

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

wip

wip

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

wip

wip

wip

wip

wip

WIP

WIP

WIP

wip

fixes

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

wip

wip

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

[WIP] Migrate tuple(handle) -> handle

wip

wip

wip

wip

Tests

Tests

Tests

Feedback

Feedback

Feedback

Feedback

Feedback

test

test

test

test

Tests

Fix

tests

fixes

Simplify handle indexing

Simplify handle pre_forward_order indexing

fixes

fixes

fixes

fixes

fixes

fixes

fixes

wip

runner

fix

fix

fix

fix

wip

wip

wip

ghstack-source-id: 82e17e25f4ca9567e7c86f42be8355a86f1baabc
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106738

fixes

fixes

fixes

simplify

simplify

simplify

simplify

Fixes

fixes

Fix

Lint

Lint

rm

Gen fixes

voznesenskym committed:
shit hacks wip

shit hacks wip

Better modules

wip

wip

wip

wip

WIP

voznesenskym committed:
wip

voznesenskym committed:
wtf

voznesenskym committed:
wtf

voznesenskym committed:
wtf

voznesenskym committed:
woof

voznesenskym committed:
woof

voznesenskym committed:
WIP

voznesenskym committed:
Store dynamo hook registrations in residuals

ghstack-source-id: bf34bc9bcbd4306d8459b1b53707a359a48f5a41
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107502

Store dynamo hook registrations in residuals

Store dynamo hook registrations in residuals

lint

lint

voznesenskym committed:
Fixes

voznesenskym committed:
WIP

voznesenskym committed:
Wip

voznesenskym committed:
Wip

voznesenskym committed:
Wip

voznesenskym committed:
wip

voznesenskym committed:
WIP

voznesenskym committed:
WIP

voznesenskym committed:
WIP

voznesenskym committed:
wip

voznesenskym committed:
wip

voznesenskym committed:
WIP

voznesenskym committed:
WIP

voznesenskym committed:
wip

voznesenskym committed:
wip

voznesenskym committed:
wip

voznesenskym committed:
wip

voznesenskym committed:
wip

Commit:
wip

Commit:
wip

voznesenskym committed:
Store dynamo hook registrations in residuals

Store dynamo hook registrations in residuals

Store dynamo hook registrations in residuals

lint

lint

Fixes

WIP

Improvements

Inductor input exclusions

test crap

simpler

Fixes

Fixes

Fixes

Fixes

wip

works

works

Feedback

wip

wip

wip

wip

wip

wip

wip

wip

wip

wip

Better hook logic

Better hook logic

Better hook logic

Better hook logic

Better hook logic

Better hook logic

Fix

Fix

Fix

ghstack-source-id: 961a856a126dc7364934410abe9366b4829a5828
Pull Request resolved: https://github.com/pytorch/pytorch/pull/109537

Fix

Fix

Fix

lint

lint

lint

Fix breaks

Fix breaks

Fix breaks

Fix breaks

Fix breaks

Fix breaks

Fix breaks

Comments

voznesenskym committed:
BS conflict

voznesenskym committed:
BS conflict

voznesenskym committed:
BS conflict

voznesenskym committed:
BS conflict

voznesenskym committed:
fix rebase

voznesenskym committed:
fsdp

voznesenskym committed:
fsdp

voznesenskym committed:
fsdp

voznesenskym committed:
fsdp

voznesenskym committed:
fuck it

voznesenskym committed:
Fuck it

voznesenskym committed:
poo

voznesenskym committed:
Works eager

voznesenskym committed:
Works eager

voznesenskym committed:
wip

voznesenskym committed:
Fixes

ezyang committed:
delete some unnecessary jitter

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
restore autograd invariant testing

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
make TARGET logging

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Logify the rest of the relevant stuff

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
annotate top level exceptions with rank

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Turn off mutation in aot autograd

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

voznesenskym committed:
Remove cross device

voznesenskym committed:
Lint

voznesenskym committed:
Remove some spurious stuff

ezyang committed:
c++ log rank too

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
format_node_compact

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
misc logging, needs torchviz

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
snapshot

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
snapshot

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
checkpoint logging improvements

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
more logging checkpoint

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
logging fixes

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Switch defaults on fsdp.py

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -995,12 +995,10 @@ def source_ref(source):
         if "G" in global_builder.scope:
             local_builder.scope["G"] = global_builder.scope["G"]
+        if "L" in local_builder.scope:
+            global_builder.scope["L"] = local_builder.scope["L"]
```
What's this doing?

voznesenskym replied:
Papering over something with a hack

ezyang commented on torch/_dynamo/allowed_functions.py:
```diff
@@ -309,7 +309,40 @@ def _builtin_constant_ids():
     return rv
 
 
+# A subcheck of is_allowed, we utilize this for patching is_allowed around distributed.
```
@yanboliang curious how you would rewrite this code after your refactor

yanboliang replied:
After reading the comments here, it seems we are trying to inline these functions rather than put them as FX graph node. After my refactor, these functions should __NOT in the allow function list__ and should __in the force inline list__. I think it should be easy to add them to the right list if my refactor really works well.
The principle of my refactor is to make every function __ONLY__ fall into one of the three: allow_in_graph, inline, skip.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1806,6 +1832,7 @@ def merge_view_inputs(
 ) -> Tuple[List[Any], Optional[List[Union[int, Tuple[int, torch.Tensor]]]]]:
     assert len(fwd_inputs) == len(mutated_input_info)
     storage_ref_to_idx: Dict[StorageWeakRef, List[int]] = collections.defaultdict(list)
+    return fwd_inputs, None
```
🤔

voznesenskym replied:
This is a Certified @bdhirsh + I plan lol.

This is known from the talk as "Handling mutations on an input that aliases other inputs"

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -769,7 +778,7 @@ def inner(*flat_args):
                 torch._sync(f_arg)
                 new_arg = torch._from_functional_tensor(f_arg)
-            if arg is not new_arg:
+            if arg is not new_arg and False:
```
forces aot autograd to treat the function as not mutating inputs

ezyang committed:
Delete logging that will impede dynamo

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Cherry-pick jansel https://github.com/pytorch/pytorch/pull/110884/files

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Temporarily disable post backward final callback

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

voznesenskym committed:
Set fixes, post rebase fixes

voznesenskym committed:
distributed skipfile and allow check

[ghstack-poisoned]

voznesenskym committed:
Fix

jansel committed:
[Compiled Autograd] Fix size mismatch with FDSP resizing

voznesenskym committed:
BS

voznesenskym committed:
Munge source, merge fixes, guards, allows

jansel committed:
Jansel ops apply

jansel committed:
[Compiled Autograd] Turn accumulate_grad into an op

ghstack-source-id: e07aa9d347b364d50e9f00f2e00ccc70d23f61d9
Pull Request resolved: https://github.com/pytorch/pytorch/pull/111271

jansel committed:
[Compiled Autograd] Error if tensor_post_acc_grad_hooks is set

ghstack-source-id: 5fb7f25fa9daa3acfa24d49081ada77abb7ae7d4
Pull Request resolved: https://github.com/pytorch/pytorch/pull/111273

jansel committed:
[Compiled Autograd] Turn accumulate_grad into an op

ghstack-source-id: e07aa9d347b364d50e9f00f2e00ccc70d23f61d9
Pull Request resolved: https://github.com/pytorch/pytorch/pull/111271

jansel committed:
[inductor] Move inductor ops to CompositeExplicitAutograd

ghstack-source-id: f1501986717043d447c86f7f0ec75140dfbb2c17
Pull Request resolved: https://github.com/pytorch/pytorch/pull/111274

voznesenskym committed:
fixes

voznesenskym committed:
Revert "[Compiled Autograd] Fix size mismatch with FDSP resizing"

This reverts commit 8ef8557e5e1ae24fdb0f19765bb89815e33c5979.

voznesenskym committed:
Revert jansel, fix stuff

Commit:
rm name log

voznesenskym committed:
fix sled

voznesenskym committed:
[fsdp] Replace acc_grad hooking with register_post_accumulate_grad_hook on flat_param

voznesenskym committed:
hacks wip

voznesenskym committed:
Not working, hooks firing

voznesenskym committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into voz/fsdp_autograd3

voznesenskym committed:
rebase fixes

voznesenskym committed:
CA fix

voznesenskym committed:
cleanups, maybe bad

voznesenskym committed:
cleanup

voznesenskym committed:
[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph

ghstack-source-id: fd0b3fa7e5eec9724574946179030dc811eb5419
Pull Request resolved: https://github.com/pytorch/pytorch/pull/112811

[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph

[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph

[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph

[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph

[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph

[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph

voznesenskym committed:
lint

voznesenskym committed:
lint

voznesenskym committed:
lint

voznesenskym committed:
hacks

voznesenskym committed:
Fix grad

voznesenskym committed:
Fix grad

voznesenskym committed:
Add KnownCreateSymbolicPolicy

Wire up fake

Fake mode stuff

ghstack-source-id: 686fc9fc85f43b4d6effc5dd0993824f40d6ed03
Pull Request resolved: https://github.com/pytorch/pytorch/pull/113926

no policy for static

Refactor

Refactor

Refactor

Refactor

Refactor

Fix fake setting

Fixes

typo

not public

voznesenskym committed:
Merge commit '855a5cf42786e73e3f91d1acf73645ee283b70a2' into HEAD

voznesenskym committed:
WIP merge

voznesenskym committed:
wip

voznesenskym committed:
It works now

voznesenskym committed:
name

voznesenskym committed:
wip

voznesenskym committed:
fixes

voznesenskym committed:
Fix

voznesenskym committed:
rm

voznesenskym committed:
rm

voznesenskym committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into voz/fsdp_autograd3

voznesenskym committed:
rm

voznesenskym committed:
Rm

voznesenskym committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into voz/fsdp_autograd3

voznesenskym committed:
fixes

voznesenskym committed:
rm stuff

voznesenskym committed:
Fixes - streams

voznesenskym committed:
undo fsdp crap

voznesenskym committed:
uncomment stream stuff

voznesenskym committed:
fixes - streams

voznesenskym committed:
Fix named buffers

voznesenskym committed:
fixes

voznesenskym committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into voz/fsdp_autograd3

voznesenskym committed:
Fixups

voznesenskym committed:
More cleanup

voznesenskym committed:
Remove more

voznesenskym committed:
Fixes aot hacks

voznesenskym committed:
Fixes

voznesenskym committed:
rm

voznesenskym committed:
Rm

voznesenskym committed:
Fix counter

voznesenskym committed:
fixes

voznesenskym committed:
alt

voznesenskym committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into where_am_i

voznesenskym committed:
rebase, fixes

voznesenskym committed:
Fixes

voznesenskym committed:
rm

voznesenskym committed:
Fixes
